### ISSUE TITLE
Documentation (in particular Observable.webSocket)

### ISSUE THREAD

**Issue by timjacobi** [ReactiveX/rxjs#2193]
Hi Guys,

First of all congrats on releasing the final version, I appreciate the work everyone has put into this project as I've been closely following. Now that the final version is released I was hoping to find some more complete API documentation. In this case I was looking to create an observable from web sockets. 

First I went to the [API docs](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-webSocket) which tell me I need to pass `urlConfigOrSource` but I could not figure out how to set it up, am I missing something? 
Moving on `Observable.webSocket` returns a `WebSocketSubject` so I tried to find out how to use it. All the API docs tell me is _We need this JSDoc comment for affecting ESDoc._

Next I attempted to find some docs in this repo so I went [here](https://github.com/ReactiveX/rxjs/blob/master/doc/index.md) and clicked on [Full Reference](» Full reference) but got a 404.

I would really like to fully start using RxJS 5 when we build our next product but the lack of documentation prevents us from doing so. If we were to use it then I'm sure we would also contribute if we find bugs as we go along.

Am I missing an important piece here?
Are there any examples elsewhere?
How can I get involved to help improving the docs?

Again thanks for this great library.

**Comment by kwonoj** [comment]
Yes, while it is released version it still lacks some of the documentation in detail, need to be improved over time. If you're using TypeScript in codebases, most of type comes with the inference that you can get it immediately (i.e `config` interface you mentioned: https://github.com/ReactiveX/rxjs/blob/master/src/observable/dom/WebSocketSubject.ts#L13) but that isn't applicable for all cases since it's TypeScript only.

For now, if you find some of docs are missing and looking for concise examples of code snippet, unit test in repo provides immediate workarounds. For example, you can find several approach of constructing websocketsubject in test case (https://github.com/ReactiveX/rxjs/blob/master/spec/observables/dom/webSocket-spec.ts#L226) with its usages. 

It is hugely appreciated to have contribution to improve document - first you can check general contribution guideline here (https://github.com/ReactiveX/rxjs/blob/master/CONTRIBUTING.md) and use other documented codebase as reference (https://github.com/ReactiveX/rxjs/blob/master/src/operator/filter.ts for example).

**Comment by blesh** [comment]
The bottom line here is that we need to add documentation around WebSocketSubject. Thanks @timjacobi for bringing this back to our attention. I think we should prioritize documentation as well.

**Comment by timjacobi** [comment]
@kwonoj Thanks for pointing out the tests, I was actually able to infer a lot of stuff from there 👍

**Comment by jrgleason** [comment]
It would be nice if we could add more details about how to test this too. I looked at https://github.com/ReactiveX/rxjs/blob/9b428d7778ae8ac35c729e66a238ce97698a252e/spec/observables/dom/webSocket-spec.ts but I couldn't really figure it out.

**Comment by marco-fp** [comment]
I can't find any documentation in regards to the use of webSocket in v6, is this a work in progress?

**Comment by luillyfe** [comment]
Hi guys, any advance in this topic ? because I got a proposal, it is pretty straightforward (a news feed) but it shows a real use case where we can use websocket subject.

https://github.com/luillyfe/rxjs-websockets/tree/features/news-feed.

Reviews so Welcome!

**Comment by luillyfe** [comment]
Here I come with Multiplexing example (It has News and Chat),

https://github.com/luillyfe/rxjs-websockets/tree/features/multiplexing

I want to hear your opinions.

**Comment by dalu** [comment]
An Angular example would be lovely, because right now I'm hitting walls everywhere.
When ever I'm trying to send something I'm getting a CloseEvent.
I have a simple chat, with active user list and chat messages, and pings so the server knows the user is active.
It works with wscat, it doesn't with Angular, rx subjects and observables, and neither with rx webSocket.
Every time I send something CloseEvent. game over. Direct connection, any browser.

**Comment by dalu** [comment]
How do you NOT send a message with multiplex on subscribe and on unsubscribe?

**Comment by luillyfe** [comment]
@dalu not sure what you mean by that. Do you mind to explain it a little better ?

**Comment by dalu** [comment]
.multiplex() has a signature

` multiplex(subMsg: () => any, unsubMsg: () => any, messageFilter: (value: T) => boolean)`

I mean that there are 3 parameters, the 1st 2 define a callback when the resource is subscribed to and the 2nd when unsubscribed.

`() => {return {type:'subscribe', model: 'whatever'}}`

even when you have `() => {}` , `undefined` gets sent over the socket.

I'm guessing it's not possible to not send anything on subscribe and unsubscribe.

I tried null and undefined as params
```ts
subj.multiplex(
null, // or undefined
undefined, // or null
filterfunc
)
```
 but then I receive errors because there is no check if the param is a function, null or undefined.

**Comment by inf3cti0n95** [comment]
@dalu in the multiplex() method, for message filter the filter function should return a Boolean value (true or false). returning True will result in message to pass and return false would filter out the message.

I suggest if you need to have a filter to have no messages to be passed, try out 
```js 
() => false
``` 

I hope this makes it clear.

**Comment by inf3cti0n95** [comment]
@dalu Oops! My bad!

**Comment by inf3cti0n95** [comment]
@dalu  But then If you do not want to have that sub and unSub methods to fire, why are you even using multiplex? Can't you just use the original webSocket with a filter method? 😅

**Comment by dalu** [comment]
In my case the server doesn't need to know if a client subscribe or unsubscribed. It's not relevant.

multiplex filters by type, it's all as it should be but I don't want the client to send notifications to the server when the client subscribes and unsubscribes.

There should be a check if un/subMsg === undefined or similar to do noop.

Right now I'm using
```
.multiplex({
() => ({}),
() => ({}),
filterfunc
})
```

which sends an empty object over the wire `{}`.
I don't want it to send anything.

**Comment by japalo** [comment]
@dalu Did you find a fix for this, or did you end up sending empty objects? I am thinking about using rxjs webSocket for a project but it feels weird to send empty objects all the time.

Is there any way of using multiplex for just listening to data being pushed instead of listening AND sending with Observable.webSocket ?

**Comment by dalu** [comment]
I didn't find a solution and went with filtering those empty objects server-side.
```go
	if string(msg) == "undefined" {
		return
	}
```
An alternative would be [EventSource](https://blog.icod.de/2018/12/17/angular-eventsource-go-and-wasted-lifetime/) for just listening to server side messages. But you have to keep it alive with pings.
Every 30s should be enough.
I'm not sure how robust websocket and especially rxjs-websocket is (think changing client IPs on mobiles, lost connections etc).
EventSource doesn't require a connection upgrade and is "essentially native http"

Of course you don't have to use the sending capabilities of a websocket and can just use it as a "listener".

**Comment by japalo** [comment]
My idea is to use a socket as a type of event listener in any case. An event from backend is triggered (with payload) and frontend responds to it depending on what type of event it is. 90% of our users are desktop users so i am kinda ok with mobile users losing connection every now and then since the main flows wont be using socket(s).

If i decide to use sockets from rxjs, will i have an issue with empty objects pushed to backed if i never trigger a `Subject.next()`?

**Comment by dalu** [comment]
only you can answer that question.

meanwhile I worked around the problem (easily) at least on the listening side, by manually filtering.
```ts
    this.subject$ = webSocket(environment.notifierUrl);
    this.threadSubject$ = this.subject$.pipe(
      filter((data: DatabaseNotification) => data.table === 'threads')
    )
```
not sure what .multiplex() does, this here works for me TM.
